src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.semconv;
+
+import io.opentelemetry.api.common.AttributeKey;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+
+public final class AttributeKeyTemplate<T> {
+
+  private final ConcurrentMap<String, AttributeKey<T>> keysCache = new ConcurrentHashMap<>();
```

Comment:
WDYT about a single static cache for all the templates?

Reply:
Yeah, and then prefix the map key with the actual prefix?

Reply:
Yeah, exactly that.

Reply:
What's the advantage of single map? Sounds like more contention and more resizing outside of calls for a particular prefix. I.e. if i call `HTTP_REQUEST_HEADER.getAttributeKey(..)` many times with distinct names but `RPC_GRPC_RESPONSE_METADATA.getAttributeKey(..)` for a small set of keys, the resizing of the shared map triggered by the cardinality of `HTTP_REQUEST_HEADER` could impact `RPC_GRPC_RESPONSE_METADATA`.

I suppose using a single map minimizes the memory footprint of initializing many maps, most of which won't get used for a particular application. But we could solve that by initializing the maps with a small initial capacity.

Reply:
I'm fine with any of the above proposed approaches.

> I suppose using a single map minimizes the memory footprint of initializing many maps, most of which won't get used for a particular application. But we could solve that by initializing the maps with a small initial capacity.

Another alternative would be to do lazy initialization of the individual maps, to reduce the impact of template attributes that are not being used.

Reply:
> With a single map/cache we would need to cast the AttributeKey objects, though, when retrieving.

this seems like a good enough reason to stick with instance field

Reply:
> > With a single map/cache we would need to cast the AttributeKey objects, though, when retrieving.
> 
> this seems like a decent reason to stick with instance field

That's a good point. I don't have strong feelings about that, I think the current version is okay too.